### PR TITLE
Add ConcurrentDictionary to allow performing multipart uploads in mul…

### PR DESCRIPTION
**Pull request by ganeshnj**
*Issue #, if available:* None

*Description of changes:* When using `AmazonS3EncryptionClientV2` in multi-threaded environment, it fails to store `UploadPartEncryptionContext` of a multipart upload, because by default `Dictionary` isn't a thread safe data structure in C#.

There exists [ConcurrentDictionary](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2?view=netcore-3.1) which is available on all target framework except `net35`. Therefore, I added custom `ConcurrentDictionary` which includes all the basic operations that are being used in the code at the moment.

*Testing:*: Able to multipart upload in multiple threads using `AmazonS3EncryptionClientV2`.

Note: All threads are working on separate input streams.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/Util/ConcurrentDictionary.cs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+ * 
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+using System.Collections.Generic;
+
+namespace AWSSDK.Extensions.S3.Encryption.Utils
```
Comment (normj) [inline src/Util/ConcurrentDictionary.cs:18]: Namespace should be Amazon not AWSSDK

File: src/AmazonS3EncryptionClientBase.cs [hunk 2]
```diff
@@ -70,8 +71,8 @@ internal AmazonS3Client S3ClientForInstructionFile
 	    }
 
         internal AmazonS3CryptoConfigurationBase S3CryptoConfig { get; set; }
-        internal readonly Dictionary<string, UploadPartEncryptionContext> CurrentMultiPartUploadKeys = new Dictionary<string, UploadPartEncryptionContext>();
-        internal readonly Dictionary<InitiateMultipartUploadRequest, UploadPartEncryptionContext> AllMultiPartUploadRequestContexts = new Dictionary<InitiateMultipartUploadRequest, UploadPartEncryptionContext>();
+        internal readonly ConcurrentDictionary<string, UploadPartEncryptionContext> CurrentMultiPartUploadKeys = new ConcurrentDictionary<string, UploadPartEncryptionContext>();
```
Comment (normj) [inline src/AmazonS3EncryptionClientBase.cs:74]: We should use the framework version of ConcurrentDictionary for every platform except .NET 3.5. That should just be a difference of swapping using statements depending on the platform.

File: src/Util/ConcurrentDictionary.cs (hunk 1 above)
Comment (ganeshnj) [reply src/Util/ConcurrentDictionary.cs]: Fixed.

File: src/AmazonS3EncryptionClientBase.cs (hunk 2 above)
Comment (ganeshnj) [reply src/AmazonS3EncryptionClientBase.cs]: Agreed, I fixed in revision 2.

**Pull request merged**